### ISSUE TITLE
APIs of forward multiple tokens given kv cache?

### ISSUE THREAD

**Issue by LiuXiaoxuanPKU** [NVIDIA/TensorRT-LLM#223]
As an example, we can generate from partial kv values, for example:
```
key: [k1, k2, k3]
value: [v1, v2, v3]
```
if it's the regular generation phase, given the new input token t4, we can get q4, and we can get the attention score by `mha([k1, k2, k3], [v1, v2, v3], q4)`.

I'm wondering in TensorRT-LLM, is it possible to forward on multiple tokens in a single forward pass? Specifically, the new input tokens are a list of tokens `[t4, t5, t6]` and we get the attention score (and eventually logits) by `mha([k1, k2, k3], [v1, v2, v3], [q4, q5, q6])`.

Thanks!

**Comment by juney-nvidia** [comment]
@LiuXiaoxuanPKU  thanks for bringing this interesting thought to us. Let me confirm my understanding, so you are asking in TensorRT-LLM whether it is possible to generate multiple output tokens in a single decoder generation step, right? 

Thanks
June

**Comment by LiuXiaoxuanPKU** [comment]
Thanks for the reply. Yeah, maybe more precisely, generate the logits for multiple tokens in a single decoder generation step. Still the example above, it will still only generate one token `t7`, but it will generate the logits of `[t4, t5, t6]` at the same time.

**Comment by juney-nvidia** [comment]
Interesting idea, currently TensorRT-LLM doesn't directly support such scenario, although I don't see any fundamental blockers of implementing this in TensorRT-LLM since both the attention logics and the generation logics can be customized with enough flexibility.  May I know your concrete thoughts as to the multi logits generation process in a single decoder generation step?  For example, 

```
past_key: [k1, k2, k3]
past_val: [v1, v2, v3]
the 1st generated token:
   t4
the second generation step:       
   logits generation for t5
      mha([k1, k2, k3], [v1, v2, v3], q4)
   logits generation for t6
      mha([k1, k2, k3], [v1, v2, v3], (q4, q5))
   logits for t7 and the generated token t7
      mha([k1, k2, k3], [v1, v2, v3], (q4, q5, q6))
```
For sure the above multi-steps for t5/t6/t7 can be merged as a single step if the dependency can be eliminated with some modeling considerations. 

Is this one of the possible way of implementing what you ask here? 

And out of curiosity, it looks that in this way more computations are used with only generating a single token, may I know the consideration of doing this? In case it may relate to some of your ongoing research idea, pls ignore my question to make you feel comfortable. Also, if you have interest to try with TensorRT-LLM for some research work, we are happy to have a dedicated discussion with you to see what we may help. 

June

**Comment by LiuXiaoxuanPKU** [comment]
1. Yes, exactly. 

2. It would be very helpful if the tensorRT-LLM team could help and guide the implementation. I could not find your email, please email me (xiaoxuan_liu@berkeley.edu) to set up a dedicated discussion. Appreciate it.

Lily

**Comment by juney-nvidia** [comment]
> 1. Yes, exactly.
> 2. It would be very helpful if the tensorRT-LLM team could help and guide the implementation. I could not find your email, please email me ([xiaoxuan_liu@berkeley.edu](mailto:xiaoxuan_liu@berkeley.edu)) to set up a dedicated discussion. Appreciate it.
> 
> Lily

Sure, already sent out an email to you and pls check it.

June

**Comment by 0xbitches** [comment]
This would be quite interesting to our serving architecture as well. Looks like the issue has been triaged, is an update coming soon?

**Comment by juney-nvidia** [comment]
@0xbitches 

There was some discussions with @LiuXiaoxuanPKU , although no further feedback from her recently:)

Going back to your question, there is already Python binding for TRT-LLM C++ runtime, so you can have a try with current TRT-LLM(based on the main branch) workflow to see whether it is flexible enough to add the features you want. If you feel that there is inflexibility preventing you from adding new features, pls let us know.
Thanks
June

**Comment by nv-guomingz** [comment]
do u still have further issue or question now? If not, we'll close it soon.
